Daiataxis template for documentation is ready. (#33)
FAIRmat-NFDI/pynxtools-stm: docs/explanation/reader-explanation.md

# STS Reader
The prime purpose of the reader is to transform data from measurement files into community-defined concepts constructed by the SPM community which allows experimentalists to store, organize, search, analyze, and share experimental data (only within the [NOMAD](https://nomad-lab.eu/nomad-lab/) platform) among the scientific communities. 

## Supproted Files and File Versions
- Reader builds on [NXsts](https://github.com/FAIRmat-NFDI/nexus_definitions/blob/fairmat/contributed_definitions/NXsts.nxdl.xml) application definition

> - Needs an experimental file, a config file and a eln file

Author: Resolved!

- Can parse Scanning Tunneling Spectroscopy (STS) from
    - Nanonis: Generic 5e, Generic 4.5
- Can parse Scanning Tunneling Microscopy (STM) from
    - Nanonis: Generic 5e, Generic 4.5

## NeXus Application Definition

> To define a standarized schema, we choose NeXus ontology and we defined a application definition `NXsts` for reading and standarizing `STM` as well as `STS` experimental data. To get the application definition and other related base classes please follow NeXus-FAIRmat page for [Scanning Tunneling Spectroscopy](https://fairmat-nfdi.github.io/nexus_definitions/sts-structure.html). There you will get application definition `NXsts` and other base classed used to construct the application defintion.

Author: Resolved!

## Introduction to the Different Input Files

> To utilize or reuse or extend the reader, motivation behind the reader input files must be understood. The files are using specific semantics rules so that reader can understand the files and work with the content of files.

Reviewer: ```suggestion
To utilize, reuse, or extend the reader, the different reader input files must be understood. The files are using specific semantic rules so that reader can understand the files and work with their content.
```

Author: Resolved!

The input files are:

> 1. Raw File that contains experimental data

Reviewer: ```suggestion
1. Raw File(s) containing data from experiments
```

2. ELN (Electronic Lab Notebook) to collect user input data

> 3. Config files that connects the data to NeXus application concept.

Reviewer: ```suggestion
3. Config files that connect the data to concepts in the NeXus application definition `NXsts`.
```

Author: Resolved!

### Raw File

> This type of file (such as `example.dat` or `example.sxm`) just comes from the experiment. 

Reviewer: ```suggestion
This type of file (such as `example.dat` or `example.sxm`) is the data file generated during the experiment. 
```

Author: Resolved!

The example given below is to explain the **command line ELN**.
```yaml
Instrument:
  hardware:
    name: Nanonis
  lock_in:
    lock_in_data_flip_number: -1.0
    modulation_amplitude:
      unit: V
      value: 0.005
    modulation_frequency:
      unit: Hz
      value: 973.0
    status: 'OFF'
  piezo_config:
    active_calib: LHe
  sample_bias:
    bias:
      unit: V
      value: 0.005
    bias_calibration: 1.0
    bias_offset:
      unit: V
      value: '0'
  software:
    rt_release: '10771'
    ui_release: '10771'
    vendor: nanonis
    version: Generic 5e
  stm_head_temp:
    unit: K
    value: 5.04866
collection_identifier: TiSe2_2303a_annealing_300C_5min_evaporate_Pyrene_1_
default: backward
definition: NXsts
entry_identifier: TiSe2_2303a_annealing_300C_5min_evaporate_Pyrene_1_0070
experiment_description: A new TiSe2, annealed at 300 C for 5 min, then cool down to
  RT, evaporate the Pyrene on RT, 2.2 E -7, totally 10 s.
experiment_identifier: C:\Users\SPM-PEEM\Desktop\DATA_Nanonis\20220711_CreaTec_Service_Benchmarks_LHe\Nanonis-Session-PMD100-HVHU_CreaTec_Service_PalmaLabBerlin220711
experiment_type: stm
type: background
```

> This type of ELN needs to be used if the reader is run from the command line. To know what fields and groups refer which type of data, one needs to read the NeXus definition on the [FAIRmat NeXus Proposal](https://fairmat-nfdi.github.io/nexus_definitions/classes/contributed_definitions/NXsts.html#nxsts) page or in the [GitHub repository](https://github.com/FAIRmat-NFDI/nexus_definitions/blob/fairmat/contributed_definitions/NXsts.nxdl.xml). 

Reviewer: ```suggestion
This type of ELN needs to be used if the reader is run from the command line. To know which fields and groups refer to which type of data, one needs to read the NeXus definition on the [FAIRmat NeXus Proposal](https://fairmat-nfdi.github.io/nexus_definitions/classes/contributed_definitions/NXsts.html#nxsts) page or in the [GitHub repository](https://github.com/FAIRmat-NFDI/nexus_definitions/blob/fairmat/contributed_definitions/NXsts.nxdl.xml). 
```

Author: Resolved!

                                   "/dat_mat_components/Temperature 1 [filt]/metadata",
                                   "/dat_mat_components/LI Demod 1 X [filt]/value",
                                   "/dat_mat_components/LI Demod 1 X [filt]/unit",
                                   "/dat_mat_components/LI Demod 1 X [filt]/metadata",
                                   "/dat_mat_components/LI Demod 1 Y [filt]/value",
                                   "/dat_mat_components/LI Demod 1 Y [filt]/unit",
                                   "/dat_mat_components/LI Demod 1 Y [filt]/metadata",
                                   "/dat_mat_components/LI Demod 2 X [filt]/value",
                                   "/dat_mat_components/LI Demod 2 X [filt]/unit",
                                   "/dat_mat_components/LI Demod 2 X [filt]/metadata",
                                   "/dat_mat_components/LI Demod 2 Y [filt]/value",
                                   "/dat_mat_components/LI Demod 2 Y [filt]/unit",
                                   "/dat_mat_components/LI Demod 2 Y [filt]/metadata"]}
}
```
**NOTES**

> - Each key is nevigating the NeXus concept (e.g. `/ENTRY[entry]/INSTRUMENT[instrument]/piezo_config/active_calib` key nevigates `ENTRY` -> `INSTRUMENT` -> `piezo_config` -> `active_calib` field in `NXsts` application definition.) in NeXus application definition.  
> - Value denoted by `@eln` refers the data must come from eln (user provided), but can be changed if rwa file contains that piece of data. 

Reviewer: ```suggestion
- Each key is pointing to the NeXus concept (e.g. `/ENTRY[entry]/INSTRUMENT[instrument]/piezo_config/active_calib` key nevigates `ENTRY` -> `INSTRUMENT` -> `piezo_config` -> `active_calib` field in `NXsts` application definition.) in the NeXus application definition.  
- If the value is denoted by the token `@eln`, the data must come from the ELN (user provided), but this can be changed if the raw file contains that piece of data as well. 
```

Author: Resolved!

- To update (if needed) the config file, a set of rules needs to be followed:

>   - The dictionary in config files have the following meaning:

Reviewer: ```suggestion
  - The dictionaries in the config files have the following meaning:
```

Author: Resolved!

    ```
    "/ENTRY[entry]/INSTRUMENT[instrument]/lock_in/harmonic_order_N": {"D1": {"value": "/Lock-in/Harmonic D1/value"},
                                                                      "D2": {"value": "/Lock-in/Harmonic D2/value"}},
    ```

>     Here, the part `N` in field `harmonic_order_N`, can be considered as the name of dimensions, can be replaced by `D1` and `D2` to  write two `harmonic_order` and can be extended for further dimensions.

Reviewer: ```suggestion
    Here, the part `N` in field `harmonic_order_N` can be considered as the name of dimensions and can be replaced by `D1` and `D2` to  write two fields of `harmonic_order` . This can be extended to further dimensions.
```

Author: Resolved!

  - List for the same concept
    ```
    "/ENTRY[entry]/INSTRUMENT[instrument]/piezo_config/active_calib": ["/Piezo Configuration/Active Calib.",
                                                                       "/Piezo Calibration/Active Calib."],
    ```
    For different type of software versions the raw data path could be different for the same concept. For example, Nanonis software `generic 5e` has `/Piezo Configuration/Active Calib.` and generic 4.5 has `/Piezo Calibration/Active Calib.` for the same concept `/ENTRY[entry]/INSTRUMENT[instrument]/piezo_config/active_calib`.

>   - In the config file, the concept that takes data from ELN is denoted by `@eln`. Otherwise, data will come from experimental raw files.

Reviewer: ```suggestion
  - In the config file, concepts that take data from the ELN are denoted by `@eln`. Otherwise, data will come from experimental raw files.
```

Author: Resolved!

>   - Importantly one should note that the `NXdata` concept `/ENTRY[entry]/DATA[data]` takes a dict of lists. Each key (`0`, `1` ...) of the dict refers to an NXdata group of with fields `bias` and `current` for multiple given setups, i.e, with and without `filter` check points. For another setup, one can extend the dict following the same convention used here.

Reviewer: ```suggestion
  - Importantly, the `NXdata` concept `/ENTRY[entry]/DATA[data]` takes a dict of lists. Each key (`0`, `1` ...) of the dict refers to an NXdata group with fields `bias` and `current` for multiple given setups, i.e, with and without `filter` check points. For another setup, one can extend the dict following the same convention used here.
```

Author: Resolved!

## Useful Functions:
There are a few functions that you can utilize to make this reader compatible with your data:

> - **Function get_stm_raw_file_info()**: For `STM` experiment the function can return you the slash separated dict in a text file. This dict helps to write or modify the config file according to your raw data file. 

Reviewer: ```suggestion
- **Function `get_stm_raw_file_info()`**: For `STM` experiments, the function can return the slash separated dict in a text file. This dict helps to write or modify the config file according to the raw data file. 
```

Author: Resolved!
